KAFKA-13036: Replace EasyMock and PowerMock with Mockito for RocksDBMetricsRecorderTest

divijvaidya opened the pull request:
## Changes
1. Migrate to Mockito
2. Add more assertive checks using `verify`
3. Minor indentation fixes

## Testing
`./gradlew streams:unitTest --tests RocksDBMetricsRecorderTest` is successful

## Notes to help the reviewers
1. Mockito uses `MockitoJUnitRunner.StrictStubs.class` to assert that all stubs are verified.
2. Mockito does not have the concept of "register-activate" represented by "when - replay" in EasyMock. Hence, you only need to use "when" and that is all. No "replay" is required.
3. `PowerMock.reset` is replaced by `Mockito.reset`
4. Mockito does not have to distinguish between "niceMock" and a normal "mock". All mocks are "nicemocks".
5. Mockito can mock static methods using `mockStatic`
6. When `verify(mockObj).functionName()` is called, it is equivalent to `verify(mockObj, times(1)).functionName()` i.e. it validates that the function in invoked on mocked object exactly once.

## Code coverage
### Before
![Screenshot 2022-08-02 at 18 14 26](https://user-images.githubusercontent.com/71267/182423056-b6bfdc1a-4587-434f-ba1d-9bd55e045a33.png)

### After
![Screenshot 2022-08-02 at 18 14 41](https://user-images.githubusercontent.com/71267/182423081-43091937-d77a-45e3-8119-d66aea415dad.png)

divijvaidya commented:
@ableegoldman @cadonna please help review this test in the `streams` domain.

dplavcic commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -70,105 +73,136 @@ public class RocksDBMetricsRecorderTest {
     public void shouldSetStatsLevelToExceptDetailedTimersWhenValueProvidersWithStatisticsAreAdded() {
-        statisticsToAdd1.setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);
-        replay(statisticsToAdd1);
+        doNothing().when(statisticsToAdd1).setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);
```
I'm not sure if there is a convention on when to stub things, but according to the docs this is redundant (especially in combination with the `verify` at the end):
> Use doNothing() for setting void methods to do nothing. **Beware that void methods on mocks do nothing by default!** However, there are rare situations when doNothing() comes handy: 
> - Stubbing consecutive calls on a void method..
> - When you spy real objects and you want the void method to do nothing ...

- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#when(T)

Personally, I would just leave the `verify(statisticsToAdd1).setStatsLevel(eq(StatsLevel.EXCEPT_DETAILED_TIMERS));` line we have

divijvaidya replied:
That's fair. I have removed the stubbing for void methods in the latest revision.

dplavcic commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +313,349 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+        final long now = 0L;
+        when(statisticsToAdd1.getHistogramData(any())).thenReturn(new HistogramData(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0L, 0L, 0.0));
+        when(statisticsToAdd1.getAndResetTickerCount(any())).thenReturn(0L);
+        doNothing().when(compactionTimeAvgSensor).record(eq(0d), eq(now));
```
There is no need to use matchers when all provided args are raw values.

This can be updated to `doNothing().when(compactionTimeAvgSensor).record(0d, now);`

divijvaidya replied:
I didn't know this. Thank you for pointing it out. Changed.

dplavcic commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -36,22 +38,23 @@
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
```
What do you think we use `@RunWith(MockitoJUnitRunner.StrictStubs.class)` which automatically detects stubbing argument mismatches instead of the `@RunWith(MockitoJUnitRunner.class)` 🤔 
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/junit/MockitoJUnitRunner.html
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/junit/MockitoJUnitRunner.StrictStubs.html
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/quality/Strictness.html#STRICT_STUBS

With `StrictStubs` we get (c/p from the [docs](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/quality/Strictness.html#STRICT_STUBS)):
> - Improved productivity: the test fails early when code under test invokes stubbed method with different arguments (see [PotentialStubbingProblem](https://javadoc.io/static/org.mockito/mockito-core/4.6.1/org/mockito/exceptions/misusing/PotentialStubbingProblem.html)).
> - Cleaner tests without unnecessary stubbings: the test fails when unused stubs are present (see [UnnecessaryStubbingException](https://javadoc.io/static/org.mockito/mockito-core/4.6.1/org/mockito/exceptions/misusing/UnnecessaryStubbingException.html)).
> - Cleaner, more DRY tests ("Don't Repeat Yourself"): If you use [Mockito.verifyNoMoreInteractions(Object...)](https://javadoc.io/static/org.mockito/mockito-core/4.6.1/org/mockito/Mockito.html#verifyNoMoreInteractions(java.lang.Object...)) you no longer need to explicitly verify stubbed invocations. They are automatically verified for you.

Note that once we migrate this test to use JUnit5, and we start using `@ExtendWith(MockitoExtension.class)` instead of the `@RunWith(MockitoJUnitRunner.class)` we will get `STRICT_STUBS` behavior out-of-the-box.

---

Considering this test for example:
```java
    @Test
    public void shouldSetStatsLevelToExceptDetailedTimersWhenValueProvidersWithStatisticsAreAdded() {
        doNothing().when(statisticsToAdd1).setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);

        recorder.addValueProviders(SEGMENT_STORE_NAME_1, dbToAdd1, cacheToAdd1, statisticsToAdd1);

        verify(statisticsToAdd1).setStatsLevel(eq(StatsLevel.EXCEPT_DETAILED_TIMERS));
    }
```

- with `StrictStubs`, if someone changes what is passed to the `setStatsLevel()` method in production code, and hard-code it to, for example `statistics.setStatsLevel(StatsLevel.ALL);`, this test would fail **even without explicit verify** with following message:

```
Unnecessary stubbings detected in test class: RocksDBMetricsRecorderTest
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at org.apache.kafka.streams.state.internals.metrics.RocksDBMetricsRecorderTest.shouldSetStatsLevelToExceptDetailedTimersWhenValueProvidersWithStatisticsAreAdded(RocksDBMetricsRecorderTest.java:180)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
...
```

Mockito documentation provides a good explanation of what is happening (and why verify step could be redundant):
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#when(T)
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#verify(T)
> Although it is possible to verify a stubbed invocation, usually it's just redundant. Let's say you've stubbed foo.bar(). If your code cares what foo.bar() returns then something else breaks(often before even verify() gets executed). If your code doesn't care what get(0) returns then it should not be stubbed. 


Note:
- I'm not sure if Mockito docs links with `fragment/anchor` don't work just for me or for all 🤔

divijvaidya replied:
> @RunWith(MockitoJUnitRunner.StrictStubs.class)

I explicitly added this, though, shouldn't it be default behaviour already since 3.x version of Mockito (we are on 4.4)?

> this test would fail even without explicit verify with following message:

That is fair. We can either specify the expectation using a stub and no verify OR we can avoid the stud for void method and verify the call. I opted for latter in latest revision.

dplavcic replied:
> I explicitly added this, though, shouldn't it be default behaviour already since 3.x version of Mockito (we are on 4.4)?

You're right! It is the default behavior and it works as expected and described in the docs. It's not needed to use `@RunWith(MockitoJUnitRunner.StrictStubs.class)`. I apologize.

Before adding the initial comment about `StrictStubs` usage I double checked the docs, and I have updated few test cases with stubs that are never invoked, and tests were still passing 😕 Only once I added explicit `StrictStubs` tests started to fail (that's the reasoning behind initial comment, but it seems something wasn't working as expected on my machine 😅 )

divijvaidya replied:
You were not completely wrong. I have found the following idiosyncrasies:
1. `StrictStubs` does not verify invocation of mocks for `mockstatic` cases. You need an explicit `verify` call to do that.
2. When you run a test `StrictStubs` does not perform verification but when you run all together, it does ensure that all stubs are being invoked.

I think it's safer to keep `StrictStubs` for now, so let's keep the change as you suggested.

dplavcic commented:
Regarding "Notes to help the reviewers":
> 1. Mockito does not have `verifyAll()` hence each mocked function call needs to be verified individually.

This depends on the `strictness` level used while creating mocks.
Please check:
-  https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/quality/Strictness.html#STRICT_STUBS
- https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#mockito_strictness

divijvaidya commented:
@dplavcic Thank you for your review and enlightening me on better usage of Mockito. I have addressed your comments. Please take another look when you get a chance.

dplavcic commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -70,105 +73,133 @@ public class RocksDBMetricsRecorderTest {
         recorder.addValueProviders(SEGMENT_STORE_NAME_1, dbToAdd1, cacheToAdd1, statisticsToAdd1);
-
-        verify(statisticsToAdd1);
+        verify(statisticsToAdd1).setStatsLevel(eq(StatsLevel.EXCEPT_DETAILED_TIMERS));
```
`eq` matcher is not needed here.

divijvaidya replied:
Thanks for catching. Fixed.

dplavcic commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
         recorder.addValueProviders(SEGMENT_STORE_NAME_1, dbToAdd1, cacheToAdd1, statisticsToAdd1);
 
-        verify(recordingTrigger);
+        verify(recordingTrigger).addMetricsRecorder(eq(recorder));
```
`eq` matcher is not needed here.

divijvaidya replied:
Fixed

dplavcic commented:
Just one more question. Do you maybe have `jacoco` test coverage reports before and after these changes? I think `jacoco` reports would be useful to confirm we are still covering (at least) the same amount of production code..

divijvaidya commented:
@dplavcic I have added the test coverage report to the description of this ticket. This revision is ready for your review now. Thanks again for your patience and spending time in reviewing this code. I appreciate it very much!

dplavcic commented:
Looks good to me.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -70,105 +73,133 @@ public class RocksDBMetricsRecorderTest {
         assertThrows(
-            IllegalStateException.class,
-            () -> recorder.init(streamsMetrics, TASK_ID2)
+                IllegalStateException.class,
+                () -> recorder.init(streamsMetrics, TASK_ID2)
```
nit: The indentation was actually correct before. We use 4 spaces to keep the code a bit more compact, the code is not 100% consistent with rule, though. Same applies to other indentations in this file.

divijvaidya replied:
Fixed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -70,105 +73,133 @@ public class RocksDBMetricsRecorderTest {
 
+    @After
+    public void cleanUpMocks() {
+        dbMetrics.close();
```
The Mockito docs say:

> To make sure a static mock remains temporary, it is recommended to define the scope within a try-with-resources construct.

https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#static_mocks

So I think it would make the code more robust if we used a try-with-resource construct.

Test method `shouldSetStatsLevelToExceptDetailedTimersWhenValueProvidersWithStatisticsAreAdded()` would become

```
  @Test
  public void shouldSetStatsLevelToExceptDetailedTimersWhenValueProvidersWithStatisticsAreAdded() {
      final RocksDBMetricContext metricsContext =
          new RocksDBMetricContext(TASK_ID1.toString(), METRICS_SCOPE, STORE_NAME);
      try (final MockedStatic<RocksDBMetrics> dbMetrics = mockStatic(RocksDBMetrics.class)) {
          setUpMetricsMock(dbMetrics, metricsContext);
          recorder.init(streamsMetrics, TASK_ID1);
          recorder.addValueProviders(SEGMENT_STORE_NAME_1, dbToAdd1, cacheToAdd1, statisticsToAdd1);
          verify(statisticsToAdd1).setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);
      }
  }
```
where `metricsContext` might probably be specified as a test object field depedent on the other test methods.

divijvaidya replied:
The current change is robust as well since we are closing the object in the lifecycle hook of a unit test which will guarantee that close is always called at the end of a unit test. Hence, I don't think that using try-with-resource is improving code quality here.

On the other hand, the disadvantage of using try-with-resource is that we will have to duplicate same code (or same function) for initialising `dbMetrics`, initializing mocks and calling init() locally across all unit tests, which is superfluous given that lifecycle hooks can do that you in JUnit.

What do you think? (I don't have strong opinion on this one and if you say we still want to go with try-with-resource, I will be happy to)

cadonna replied:
Originally my concern was that the lifecycle hook is that robust against refactorings that might remove the closing of the static mock. Removing a try-with-resource statement is much harder. I remember the first time we encountered the missing close of a static mock, the debugging took quite a while until we figured that we needed to close the static mock. That might be different now that we already encountered the issue once, though.

So I am fine with leaving it as you suggest with the lifecycle hook.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+        dbMetrics.when(() -> RocksDBMetrics.addEstimateTableReadersMemMetric(eq(streamsMetrics), eq(metricsContext), any()))
+                .thenAnswer(invocation -> null);
+        dbMetrics.when(() -> RocksDBMetrics.addBackgroundErrorsMetric(eq(streamsMetrics), eq(metricsContext), any()))
+                .thenAnswer(invocation -> null);
```
I think  

```
dbMetrics.when(() -> RocksDBMetrics.addNumImmutableMemTableMetric(eq(streamsMetrics), eq(metricsContext), any())).thenAnswer(invocation -> null);
...
```
```
dbMetrics.verify(() -> RocksDBMetrics.addNumImmutableMemTableMetric(eq(streamsMetrics), eq(metricsContext), any()));
...
```

We just want to verify that these methods are called during `recorder.init()`. So you can extract this verifications to a method and only call this method in `shouldInitMetricsRecorder()`. With EasyMock these calls needed to be here since EasyMock does not allow to not verify/specify an actual call in default mode.

divijvaidya replied:
Done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
-        recordingTrigger.addMetricsRecorder(recorder);
-        replay(recordingTrigger);
+
+        Mockito.reset(recordingTrigger);
```
The use of `reset()` is discouraged (see https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#reset(T...)).
You can equally verify that the method `recordingTrigger.addMetricsRecorder(recorder)` is called twice.

divijvaidya replied:
Removed as suggested.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
     }
 
     @Test
     public void shouldNotAddItselfToRecordingTriggerWhenNotEmpty2() {
```
Could you please remove the trailing `2` from the test name? I do not know why it was added.

divijvaidya replied:
Done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
-        reset(recordingTrigger);
-        replay(recordingTrigger);
+
+        Mockito.reset(recordingTrigger);
```
To avoid `reset()` here, you can split this test method into two:
- in one test method you add two value providers, remove only one, and verify that `removeMetricsRecorder` is never called and 
- in the second test method you add two value providers, remove both, and verify that `removeMetricsRecorder` is called only once.

divijvaidya replied:
Done as suggested.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+
+        when(statisticsToAdd1.getAndResetTickerCount(TickerType.BYTES_WRITTEN)).thenReturn(1L);
+        when(statisticsToAdd2.getAndResetTickerCount(TickerType.BYTES_WRITTEN)).thenReturn(2L);
+        doNothing().when(bytesWrittenToDatabaseSensor).record(1 + 2, now);
```
I do not think this is needed. As it says, it does nothing 🙂. It also does not verify anything. Same applies to the other stubbings of void methods.

divijvaidya replied:
Removed

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+        verify(statisticsToAdd2, times(17)).getAndResetTickerCount(isA(TickerType.class));
+        verify(statisticsToAdd2, times(2)).getHistogramData(isA(HistogramType.class));
+        verify(statisticsToAdd2, times(2)).getHistogramData(isA(HistogramType.class));
+        verify(bytesWrittenToDatabaseSensor).record(isA(Double.class), eq(now));
```
This verification should be 

```
verify(bytesWrittenToDatabaseSensor).record(eq(Double.valueOf(1 + 2)), eq(now));
```

Similar applies to the other verifications. Use the values of the void stubbings above you need to remove.

divijvaidya replied:
Done, as suggested.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+
+        recorder.record(now);
+
+        verify(statisticsToAdd1, times(17)).getAndResetTickerCount(isA(TickerType.class));
+        verify(statisticsToAdd2, times(17)).getAndResetTickerCount(isA(TickerType.class));
+        verify(statisticsToAdd2, times(2)).getHistogramData(isA(HistogramType.class));
+        verify(statisticsToAdd2, times(2)).getHistogramData(isA(HistogramType.class));
```
You need to verify each `TickerType` and each `HistogramType` here. Using `isA()` is to general. I think the best thing would be to split this test method in one test method per computed metric.

dplavcic replied:
@cadonna, @divijvaidya, what do you think we just update this test method with "improved" verification, as @cadonna described [here](https://github.com/apache/kafka/pull/12459/files/e4e7549e229b2687f8491842a2352ecaf2563e5f#r944335669), and we create a new ticket/PR to split this test to multiple smaller tests (which would be cool)?

IMHO, that way we keep this PR short and simple and in the scope of the "Replace EasyMock and PowerMock with Mockito.." task 🤔

cadonna replied:
I am fine with both. However, I am not sure whether the "improved" verification vs. splitting unit tests will make too much difference.

cadonna replied:
I meant effort-wise in my previous comment.

dplavcic replied:
Regarding effort-wise, it should be the same. 

The only deficiency in splitting unit tests in this PR, whose main purpose is to replace `EasyMock` is that it makes it  a little bit harder to review (since we're introducing changes that are out-of-the scope).

cadonna replied:
I think I agree with you!

divijvaidya replied:
Thanks for the review and discussion here folks. I have tried to optimize this PR for code review friendliness and I agree that we should probably leave refactoring and other changes out of scope. I have addressed your comments Bruno. Please take a look when you get a chance.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
+        verify(blockCacheDataHitRatioSensor).record(0d, 0L);
+        verify(blockCacheIndexHitRatioSensor).record(0d, 0L);
+        verify(blockCacheFilterHitRatioSensor).record(0d, 0L);
+        verify(statisticsToAdd1, times(2)).getHistogramData(any());
+        verify(statisticsToAdd1, times(17)).getAndResetTickerCount(any());
```
Those verifications are not needed since you have already done them in `shouldRecordStatisticsBasedMetrics()`. This method tests the computation of the hit ratio and not the retrieval of the single values.

divijvaidya replied:
done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -279,449 +310,335 @@ public void shouldThrowIfDbToAddWasAlreadyAddedForOtherSegment() {
-        verify(compactionTimeAvgSensor);
+        verify(compactionTimeAvgSensor).record(0d, now);
+        verify(memtableAvgFlushTimeSensor).record(0d, now);
+        verify(statisticsToAdd1, times(2)).getHistogramData(any());
+        verify(statisticsToAdd1, times(17)).getAndResetTickerCount(any());
```
Those verifications are not needed since you have already done them in `shouldRecordStatisticsBasedMetrics()`. This method tests the computation of the average and not the retrieval of the single values.

divijvaidya replied:
done.

cadonna reviewed (commented):
Thank you for the PR, @divijvaidya !

Here my comments!

clolov commented:
Heya @cadonna and @dplavcic, Divij is taking some time off, so I will aim to provide an update to the comments and the pull request in the next couple of days.

divijvaidya committed:
Replace Powermock with Mockito for RocksDBMetricsRecorderTest

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorderTest.java:
```diff
@@ -381,157 +395,115 @@ public void shouldThrowIfValueProvidersToRemoveNotFound() {
+
+        verify(statisticsToAdd1, times(17)).getAndResetTickerCount(isA(TickerType.class));
+        verify(statisticsToAdd2, times(17)).getAndResetTickerCount(isA(TickerType.class));
+        verify(statisticsToAdd2, times(2)).getHistogramData(isA(HistogramType.class));
```
I think this should be `statisticsToAdd1` instead of `statisticsToAdd2`. Otherwise we would verify twice the same calls - here and in the next line.

divijvaidya replied:
Whoops! Good catch. I have fixed in the upcoming commit.

cadonna reviewed (commented):
Thanks @divijvaidya for the updates!

I think I found one mistake. The rest looks good to me!

divijvaidya committed:
Fix a bug

cadonna reviewed (approved):
LGTM!

Thanks for the PR, @divijvaidya !
